src/renderer/UniformSetter.js

```
@@ -0,0 +1,96 @@
+import { getUniforms } from './glUtil';
+
+let typeMap;
+
+export function makeUniformSetter(gl, program) {
+  const uniformInfo = getUniforms(gl, program);
+  const uniforms = {};
+  const needsUpload = [];
+
+  for (let { name, type, location } of uniformInfo) {
+    const uniform = {
+      type,
+      location,
+      v0: 0,
+      v1: 0,
+      v2: 0,
+      v3: 0
+    };
+
+    uniforms[name] = uniform;
+  }
+
+  function setUniform(name, v0, v1, v2, v3) {
+    const uni = uniforms[name];
+
+    if (!uni) {
```

Fail silently if uniform doesn't exist in shader. If a uniform is defined in GLSL but not actually used in some computation which renders to the screen, WebGL optimizes the shader by removing the uniform during compilation. Writing to an unused shader would error without this check. That could be useful to catch mistakes, but it's also useful to write to unused uniforms while developing shaders. Thoughts?

---

I remember we talked about a debug flag a few weeks ago, maybe errors should only be thrown in debug mode?

---

We could add an optional failSilently parameter?

---

I just added a new commit trying a new idea. The console warns (once) if a uniform doesn't exist, but the error is handled and the program keeps running. This way we can have unused uniforms in the shader during development, but we're still aware of any uniforms that are aren't being set due to a mismatch in names or whatever. This way we don't need to deal with the complexity of a debug flag. How does this sound?

---

perfect.